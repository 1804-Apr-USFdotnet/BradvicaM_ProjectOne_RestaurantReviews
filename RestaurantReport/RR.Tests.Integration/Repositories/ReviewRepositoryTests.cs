using System;
using System.Collections.Generic;
using System.Linq;
using ApprovalTests;
using ApprovalTests.Reporters;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using RR.Models;
using RR.Repositories;

namespace RR.Tests.Integration.Repositories
{
    [TestClass]
    public class ReviewRepositoryTests
    {
        private readonly RestaurantReportTestContext _testContext;
        private readonly ReviewRepository _reviewRepository;

        public ReviewRepositoryTests()
        {
            _testContext = new RestaurantReportTestContext();
            _reviewRepository = new ReviewRepository(_testContext);
        }

        [TestMethod]
        public void GetById_GivenReviewId_ReturnsCorrectReview()
        {
            var review = new Review
            {
                ReviewId = Guid.NewGuid(),
                ReviewPublicId = Guid.NewGuid(),
                Restaurant = _testContext.Restaurants.First()
            };

            _testContext.Reviews.Add(review);
            _testContext.SaveChanges();

            var result = _reviewRepository.GetById(review.ReviewPublicId);

            Assert.AreEqual(review, result);
        }

        [TestMethod]
        public void Get_OnCall_ReturnsAllReviews()
        {
            var reviews = new List<Review>
            {
                new Review
                {
                    ReviewId = Guid.NewGuid(),
                    ReviewPublicId = Guid.NewGuid(),
                    Restaurant = _testContext.Restaurants.First()
                },
                new Review
                {
                    ReviewId = Guid.NewGuid(),
                    ReviewPublicId = Guid.NewGuid(),
                    Restaurant = _testContext.Restaurants.First()
                },
                new Review
                {
                    ReviewId = Guid.NewGuid(),
                    ReviewPublicId = Guid.NewGuid(),
                    Restaurant = _testContext.Restaurants.First()
                },
            };
            
            _testContext.Reviews.RemoveRange(_testContext.Reviews);
            _testContext.Reviews.AddRange(reviews);
            _testContext.SaveChanges();

            var result = _reviewRepository.Get();

            const int expected = 3;

            Assert.AreEqual(expected, result.Count());
        }

        [TestMethod]
        public void Delete_GivenReview_RemovesReview()
        {
            var review = new Review
            {
                ReviewId = Guid.NewGuid(),
                ReviewPublicId = Guid.NewGuid(),
                Restaurant = _testContext.Restaurants.First()
            };

            _testContext.Reviews.Add(review);
            _testContext.SaveChanges();

            _reviewRepository.Delete(review);

            Assert.IsFalse(_reviewRepository.Get().Contains(review));
        }

        [TestMethod]
        [UseReporter(typeof(DiffReporter))]
        public void Update_GivenReview_UpdatesReview()
        {
            var beforeUpdate = new Review
            {
                ReviewId = Guid.Parse("4dc863f3-88da-4d3b-958c-ab585a116b9c"),
                ReviewPublicId = Guid.Parse("e5398119-bcc0-4952-a177-2d7aec86fcb5"),
                Restaurant = _testContext.Restaurants.First(),
                ReviewerName = "Mike",
                Comment = "Ok-ish",
                Rating = 5.45
            };

            _testContext.Reviews.RemoveRange(_testContext.Reviews);
            _testContext.Reviews.Add(beforeUpdate);
            _testContext.SaveChanges();

            var afterUpdate = new Review
            {
                ReviewId = beforeUpdate.ReviewId,
                ReviewerName = "John",
                Comment = "It was bad",
                Rating = 1.30,
                ReviewPublicId = beforeUpdate.ReviewPublicId,
                Restaurant = beforeUpdate.Restaurant,
                RestaurantId = beforeUpdate.RestaurantId
            };

            _reviewRepository.Update(afterUpdate);

            var result = _reviewRepository.GetById(beforeUpdate.ReviewPublicId);

            Approvals.Verify(result);
        }

        [TestMethod]
        public void Add_GivenReview_AddsReview()
        {
            var review = new Review
            {
                ReviewId = Guid.NewGuid(),
                ReviewPublicId = Guid.NewGuid(),
                Restaurant = _testContext.Restaurants.First()
            };

            _reviewRepository.Add(review);

            var reviews = _testContext.Reviews.ToList();

            Assert.IsTrue(reviews.Contains(review));
        }
    }
}
