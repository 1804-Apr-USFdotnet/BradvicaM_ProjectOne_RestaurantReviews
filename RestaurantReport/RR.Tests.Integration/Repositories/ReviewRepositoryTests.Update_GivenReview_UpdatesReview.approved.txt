
ReviewId: 4dc863f3-88da-4d3b-958c-ab585a116b9c
Id: e5398119-bcc0-4952-a177-2d7aec86fcb5
Name: John
Rating: 1.3
Comment: It was bad
RestaurantId: 3d18e4bf-768d-44ec-b3a8-02e8eec06d56
Restaurant: 
Id: 3d18e4bf-768d-44ec-b3a8-02e8eec06d56
RestaurantId: 5b2805a3-eadc-4fa0-9eba-1caab4620e07
Name: Mikes Tampa
Street: 45 Genevieve Dr.
City: Del Mar
State: CA
ZipCode: 91832
PhoneNumber: 9876253412
Website: www.mikes.com
Rating: 1.23

